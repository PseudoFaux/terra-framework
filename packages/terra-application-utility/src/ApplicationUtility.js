import ApplicationHeaderUtility from './ApplicationHeaderUtility';
import ApplicationMenuUtility from './ApplicationMenuUtility';
import UtilityUtils from './_Utils';

const ApplicationUtility = {
  ApplicationHeaderUtility,
  ApplicationMenuUtility,
  UtilityUtils,
};

export default ApplicationUtility;
export { ApplicationHeaderUtility, ApplicationMenuUtility, UtilityUtils };
