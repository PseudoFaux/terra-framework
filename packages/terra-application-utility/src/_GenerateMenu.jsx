// import React from 'react';
// import PropTypes from 'prop-types';
// import classNames from 'classnames/bind';
// import 'terra-base/lib/baseStyles';

// import Menu from 'terra-menu';

// const propTypes = {
//   /**
//    * Config File used to generate the menu
//    */
//   config: PropTypes.node.isRequired,
// };

// class GenerateMenu extends React.Component {
//   constructor(props) {
//     super(props);
//     this.generate = this.generate.bind(this);
//   }

//   componentDidMount(){
//     this.generateMenuPage(this.props.config);
//   }

//     /**
//    * Recursively traverses a config file and genereates menu pages and it's associated menu items.
//    * @param {*} config
//    */
//   generateMenuPage(config) {
//     const itemArray = [];
//     if ('children' in config) {
//       config.children.forEach((object) => {
//         const menuItem = (
//           <Menu.Item
//             key={object.key}
//             content={object.content || object.title}
//           />
//         )
//         itemArray.push(menuItem);
//         this.generateMenuPage(object);
//       });
//     } else {
//       return;
//     }
//     // console.log('menu page');
//     const menuPage = (
//       <MenuPage
//         title={config.title}
//         key={config.key}
//         onRequestBack={this.pop}
//         onRequestClose={this.push}
//         index={}
//       >
//         {itemArray}
//       </MenuPage>);
//   }
// }

// GenerateMenu.propTypes = propTypes;

// export default GenerateMenu
