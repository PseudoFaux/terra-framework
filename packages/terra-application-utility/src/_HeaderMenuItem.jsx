import React from 'react';
import PropTypes from 'prop-types';
import Arrange from 'terra-arrange';
import CheckIcon from 'terra-icon/lib/icon/IconCheckmark';
import ChevronIcon from 'terra-icon/lib/icon/IconChevronRight';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';

import styles from './MenuItem.scss';

const cx = classNames.bind(styles);

const contextTypes = {
  isGroupItem: PropTypes.bool,
  isSelectableMenu: PropTypes.bool,
};

const propTypes = {
  /**
   * Sets the item's text.
   */
  text: PropTypes.string,
  /**
   * Indicates if the item is selected. A selected item is indicated with a checkmark.
   */
  isSelected: PropTypes.bool,
  /**
   * Object containing menu item information
   */
  itemData: PropTypes.Object,
  /**
   * Unique key
   */
  key: PropTypes.string,
  /**
   * Indicates if menu item drills to a sub menu page
   */
  hasChevron: PropTypes.bool,
  // /**
  //  * List of Menu.Items to display in a submenu when this item is selected.
  //  */
  // subMenuItems: PropTypes.arrayOf(PropTypes.element),

  /**
   * Callback function for when item is clicked
   */
  onClick: PropTypes.func,

  /**
   * Callback function for when selection state changes on a selectable item.
   */
  onChange: PropTypes.func,

  /**
   * Indicates if the item has focus. This is used internally to control focus and does not set initial focus.
   */
  isActive: PropTypes.bool,
};

const defaultProps = {
  text: '',
  isSelected: false,
  isActive: false,
  isSelectable: undefined,
  isDisabled: false,
  subMenuItems: [],
};

class MenuItem extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.wrapOnClick = this.wrapOnClick.bind(this);
    this.wrapOnKeyDown = this.wrapOnKeyDown.bind(this);
    this.handleSelection = this.handleSelection.bind(this);
    this.setItemNode = this.setItemNode.bind(this);
    this.state = { isSelected: props.isSelected, key: '' };
  }

  componentDidUpdate() {
    if (this.props.isActive && this.itemNode) {
      this.itemNode.focus();
    }
  }

  setItemNode(node) {
    if (node) {
      this.itemNode = node;
    }
  }

  handleSelection(event) {
    event.preventDefault();
    const newIsSelected = !this.state.isSelected;
    this.setState({ isSelected: newIsSelected });

    if (this.props.onChange) {
      this.props.onChange(event, newIsSelected);
    }
  }

  wrapOnClick(event) {
    this.handleSelection(event);

    if (this.props.onClick) {
      this.props.onClick(event);
    }
  }

  wrapOnKeyDown(onKeyDown) {
    return ((event) => {
      if (event.nativeEvent.keyCode === MenuUtils.KEYCODES.ENTER || event.nativeEvent.keyCode === MenuUtils.KEYCODES.SPACE) {
        this.handleSelection(event);

        if (this.props.onClick) {
          this.props.onClick(event);
        }
      }

      if (onKeyDown) {
        onKeyDown(event);
      }
    });
  }

  render() {
    const {
      text,
      isSelected,
      isActive,
      ...customProps
    } = this.props;

    const { isGroupItem, isSelectableMenu } = this.context;

    const attributes = Object.assign({}, customProps);
    attributes.tabIndex = '0';

    attributes.onClick = this.wrapOnClick;
    attributes.onKeyDown = this.wrapOnKeyDown(attributes.onKeyDown);

    const itemClassNames = cx([
      'item',
      { selected: this.state.isSelected || (isGroupItem && isSelected) },
      attributes.className,
    ]);

    const textContainer = <div className={cx(['title'])}>{text}</div>;
    let content = textContainer;
    if (this.props.hasChevron || isSelectableMenu) {
      content = (
        <Arrange
          fitStart={isSelectableMenu ? <CheckIcon className={cx(['checkmark'])} /> : null}
          fill={textContainer}
          fitEnd={this.props.hasChevron ? <ChevronIcon className={cx(['chevron'])} /> : null}
          align={'center'}
        />
      );
    }

    const role = 'menuitem';

    return (
      <li {...attributes} className={itemClassNames} ref={this.setItemNode} role={role} >
        {content}
      </li>
    );
  }
}

MenuItem.propTypes = propTypes;
MenuItem.defaultProps = defaultProps;
MenuItem.contextTypes = contextTypes;

export default MenuItem;
