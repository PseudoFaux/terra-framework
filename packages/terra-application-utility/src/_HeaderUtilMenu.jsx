import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';

// import MenuItem from './MenuItem';
// import MenuItemGroup from './MenuItemGroup';
// import MenuDivider from './MenuDivider';
import Utils from './_Utils';
import MenuPage from './_HeaderUtilMenuPage';
import MenuItem from './_HeaderUtilMenuItem';

const propTypes = {
  /**
   * List of Menu.Item(s)/Menu.ItemGroup(s)/Menu.Divider(s) to be displayed as content within the Menu.
   */
  children: PropTypes.node.isRequired,
  /**
   * Callback function indicating a close condition was met, should be combined with isOpen for state management.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * CSS classnames that are append to the menu content inner.
   */
  classNameContent: PropTypes.string,
  /**
   * CSS classnames that are append to the overlay.
   */
  classNameOverlay: PropTypes.string,
  /**
   * Configuration file used to generate menu pages
   */
  menuConfig: PropTypes.object,
  /**
   * The object containing information of the user.
   */
  userData: PropTypes.element,
};

const defaultProps = {
  children: [],
  isOpen: false,
  contentWidth: '240',
};

class Menu extends React.Component {
  constructor(props) {
    super(props);
    // this.generateConfigFile = this.generateConfigFile.bind(this);
    this.setPageDimensions = this.setPageDimensions.bind(this);
    this.push = this.push.bind(this);
    this.pop = this.pop.bind(this);
    // this.generateConfig = this.generateConfig.bind(this);
    this.generateMenuPage = this.generateMenuPage.bind(this);
    this.state = { stack: [this], config: this.generateConfig() };
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.children.length !== nextProps.children.length) {
      this.setState({ stack: [this] });
    }
  }

  setPageDimensions(node) {
    if (node) {
      this.pageHeight = node.clientHeight;
      this.pageWidth = node.clientWidth;
    } else {
      this.pageHeight = undefined;
      this.pageWidth = undefined;
    }
  }

  generateConfig() {
    return {
      key: 'menu-page',
      title: 'Menu',
      children: [
        {
          key: 'user-information-page',
          title: 'User Information',
          content: this.props.userData,
          children: [
            {
              key: 'change-photo',
              title: 'User Information',
            },
          ],
        },
        {
          key: 'settings',
          title: 'Settings',
          children: [
            {
              key: 'appearance',
              title: 'Appearance',
            },
            {
              key: 'security',
              title: 'Security',
            },
          ],
        },
        {
          key: 'help-page',
          title: 'Help',
          children: [
            {
              key: 'getting-started',
              title: 'Getting Started',
            },
            {
              key: 'about',
              title: 'About',
            },
            {
              key: 'terms-of-use',
              title: 'Terms of Use',
            },
          ],
        },
      ],
    };
  }

/**
 * Recursively traverses a config file and genereates menu pages and it's associated menu items.
 * @param {*} config
 */
  generateMenuPage(config) {
    const itemArray = [];
    if ('children' in config) {
      config.children.forEach((object) => {
        const menuItem = (
          <MenuItem
            key={object.key}
            content={object.content || object.title}
          />
        )
        itemArray.push(menuItem);
        this.generateMenuPage(object);
      });
    } else {
      return;
    }
    // console.log('menu page');
    const menuPage = (
      <MenuPage
        title={config.title}
        key={config.key}
        onRequestBack={this.pop}
        onRequestClose={this.push}
        index={}
      >
        {itemArray}
      </MenuPage>);
  }

  pop() {
    if (this.state.stack.length > 1) {
      const newStack = this.state.stack.slice();
      newStack.pop();
      this.setState({ stack: newStack });
    }
  }

  push(item) {
    const newStack = this.state.stack.slice();
    newStack.push(item);
    this.setState({ stack: newStack });
  }

  render() {
    const {
      classNameContent,
      classNameOverlay,
      children,
      onChange,
      userData,
      ...customProps
    } = this.props;

    const visiblePage = this.state.stack.length - 1;
    const slides = this.state.stack.map((item, index) => (
      <MenuPage
        // eslint-disable-next-line react/no-array-index-key
        key={item.props.key}
        title={item.props.title}
        content={item.props.content}
        onRequestNext={this.push}
        onRequestBack={this.pop}
        isHidden={index !== visiblePage}
        fixedHeight={this.pageHeight}
        fixedWidth={this.pageWidth}
        refCallback={visiblePage === 0 ? this.setPageDimensions : null}
        index={index}
        isFocused={index === visiblePage}
      >
        {item.props.children}
      </MenuPage>
    ));

    return (
      <div {...customProps} >
        {slides}
      </div>
    );
  }
}

Menu.propTypes = propTypes;
Menu.defaultProps = defaultProps;
// Menu.Item = MenuItem;
// Menu.ItemGroup = MenuItemGroup;
// Menu.Divider = MenuDivider;
// Menu.Opts = {
//   widths: Popup.Opts.widths,
// };

export default Menu;
