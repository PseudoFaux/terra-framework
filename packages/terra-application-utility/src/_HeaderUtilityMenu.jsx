import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import Button from 'terra-button';
import IconLeft from 'terra-icon/lib/icon/IconLeft';
import MenuDivider from './_UtilityMenuDivider';
import Utils from './_Utils';
import MenuPage from './_HeaderUtilityMenuPage';
import styles from './_HeaderUtilityMenu.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   *
   */
  isHeightBounded: PropTypes.func,
  // /**
  //  * Key based object containing menu page data
  //  */
  // menuConfig: PropTypes.object,
  /**
   * Callback function indicating a close condition was met, should be combined with isOpen for state management.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * The base key of the pages object
   */
  rootMenuKey: PropTypes.func,
  /**
   * The object containing information of the user.
   */
  userData: PropTypes.element,
};

const defaultProps = {
  isHeightBounded: undefined,
  rootMenuKey: null,
  userData: null,
};

class HeaderUtilityMenu extends React.Component {
  constructor(props) {
    super(props);
    const map = new Map();
    this.createMap = this.createMap.bind(this);
    this.getChildren = this.getChildren.bind(this);
    this.getTitle = this.getTitle.bind(this);
    this.handleLogOut = this.handleLogOut.bind(this);
    this.handleOnChange = this.handleOnChange.bind(this);
    this.handleRequestBack = this.handleRequestBack.bind(this);
    this.pop = this.pop.bind(this);
    this.push = this.push.bind(this);
    // TODO: Remove internal generation and have consumers pass in config with additional items.
    // this.props.menuConfig = Utils.generateConfig(this.props.userData);
    // this.props.rootMenuKey = Utils.ROOTKEY;
    this.createMap(Utils.generateConfig(this.props.userData), map);
    this.state = {
      currentKey: Utils.ROOTKEY,
      previousKeyStack: [],
      map,
    };
  }

  /**
   * Return children of the specified object
   * @param {*} key
   */
  getChildren(key) {
    return this.state.map.get(key).children;
  }

  /**
   * Return the title of the specified object
   * @param {*} key
   */
  getTitle(key) {
    return this.state.map.get(key).title;
  }

  /**
   * Recursively create a map from the menu config with entries for each key.
   * A key's value is an object containing the title, content, isSelected, and children.
   * Example: {"key" => {title: "Page Title", isSelected: false, children: [{key: "child-one", title: "First Item", isSelected: false}, {key: "child-two", title: "Second Item" isSelected: false}]}
   */
  createMap(config, map) {
    map.set(
      config.key,
      { title: config.title,
        content: config.content,
        // isSelected: config.isSelected,
        children: config.children,
      },
    );
    if ('children' in config) {
      config.children.forEach((object) => {
        this.createMap(object, map);
      });
    }
  }

  handleLogOut() {
    this.props.onChange(event, 'logout');
  }

  handleOnChange(key) {
    // this.state.map.set(key).isSelected = true;
    if (!this.getChildren(key)) {
      this.props.onChange(event, key);
    } else {
      this.setState({
        previousKey: this.push(this.state.currentKey),
        currentKey: key,
      });
    }
  }

  handleRequestBack() {
    // this.state.map.set(key).isSelected = false;
    this.setState({ currentKey: this.pop() });
  }

  pop() {
    const newStack = this.state.previousKeyStack.slice();
    const lastKey = newStack.pop();
    this.setState({ previousKeyStack: newStack });
    return lastKey;
  }

  push(key) {
    const newStack = this.state.previousKeyStack.slice();
    newStack.push(key);
    this.setState({ previousKeyStack: newStack });
  }

  render() {
    const {
      onChange,
      userData,
      ...customProps
    } = this.props;

    const MenuClassNames = cx([
      'utility-menu',
      customProps.classNames,
    ]);

    const HeaderClassNames = cx([
      'header',
    ]);

    const BackButtonClassNames = cx('back-button');
    const LogOutButtonClassNames = cx('log-out-button');
    const HeaderTextClassNames = cx('header-text');

    // Display Logout button on the first page
    // Display Back button in nested pages
    const currentKey = this.state.currentKey;
    return (
      <div {...customProps} className={MenuClassNames} >
        <div className={HeaderClassNames}>
          {currentKey !== Utils.ROOTKEY && <Button className={BackButtonClassNames} onClick={this.handleRequestBack}> <IconLeft /> </Button>}
          <span className={HeaderTextClassNames}>{this.getTitle(currentKey) }</span>
        </div>
        <MenuDivider />
        <MenuPage
          pageData={this.state.map.get(currentKey)}
          onChange={this.handleOnChange}
        />
        {currentKey === Utils.ROOTKEY ? [<MenuDivider />, <Button className={LogOutButtonClassNames} onClick={this.handleLogOut}>{Utils.LOGOUT}</Button>] : null}
      </div>
    );
  }
}

HeaderUtilityMenu.propTypes = propTypes;
HeaderUtilityMenu.defaultProps = defaultProps;

export default HeaderUtilityMenu;
