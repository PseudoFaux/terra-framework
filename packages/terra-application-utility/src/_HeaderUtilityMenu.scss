:local {
  .utility-menu {
    // display: flex;
    // flex-direction: column;
    // justify-content: space-between;
    list-style-type: none;
    font-family: inherit;
    overflow: hidden;

    .header {
      align-items: center;
      display: flex;
      font-size: var(--terra-application-header-utility-menu-header-font-size, 1rem);
      font-weight: var(--terra-application-header-utility-menu-header-font-weight, bold);
      height: var(--terra-application-header-utility-menu-header-height, 2rem);

      .header-text {
        display: inline-block;
        margin-left: var(--terra-application-header-utility-menu-header-text-margin-left, 0.4286rem);
      }

      .back-button {
        background-color: var(--terra-application-header-utility-menu-header-back-button-background-color, transparent);
        border-width: var(--terra-application-header-utility-menu-header-back-button-border-width, 0);
        width: var(--terra-application-header-utility-menu-header-back-button-width, 1.714rem);
      }
    }

    .log-out-button {
      float: right;
      margin-bottom: var(--terra-application-header-utility-menu-log-out-button, 0.4286rem);
    }

  }
}
