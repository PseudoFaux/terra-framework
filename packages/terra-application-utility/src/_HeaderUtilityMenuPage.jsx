import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import Arrange from 'terra-arrange';
import IconChevronRight from 'terra-icon/lib/icon/IconChevronRight';
import MenuDivider from './_UtilityMenuDivider';
import styles from './_HeaderUtilityMenuPage.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Object containing data pertaining to this specific page
   */
  pageData: PropTypes.shape({
    Title: PropTypes.string,
    Content: PropTypes.element,
    children: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
  /**
   * Function to trigger when an item is selected
   */
  onChange: PropTypes.func.isRequired,
};

const HeaderUtilityMenuPage = ({
  pageData,
  onChange,
  ...customProps
}) => {
  const menuPageClassNames = cx('menu-page');
  const listItemClassNames = cx('list-item');
  const chevronClassNames = cx('chevron');
  // const listItemArr = pageData.children.map(child =>
  //   (
  //     <li key={child.key} className={listItemClassNames}>
  //       <Arrange
  //         onClick={() => { onChange(child.key); }}
  //         // fitStart={}
  //         fill={<div>{child.content ? child.content : child.title}</div>}
  //         fitEnd={child.children && !child.content ? <IconChevronRight className={chevronClassNames} /> : null}
  //         align={'center'}
  //       />
  //     </li>
  //   ),
  // );
  // eslint-disable jsx-a11y/no-static-element-interactions
  const listItemArr = pageData.children.map((child) => {
    if (child.content) {
      return (
        <div key={child.key}>
          <li onClick={() => { onChange(child.key); }} role="button" className={listItemClassNames}>
            {child.content}
          </li>
          { child.key === 'user-information' && <MenuDivider key={`${child.key}-divider`} />}
        </div>
      );
    }
    return (
      <li onClick={() => { onChange(child.key); }} role="button" key={child.key} className={listItemClassNames}>
        <Arrange
          // fitStart={}
          fill={<div>{child.title}</div>}
          fitEnd={child.children && !child.content ? <IconChevronRight className={chevronClassNames} /> : null}
          align={'center'}
        />
      </li>
    );
  });
// eslint-enable jsx-a11y/no-static-element-interactions
  return (
    <ul{...customProps} className={menuPageClassNames}>
      {listItemArr}
    </ul>
  );
};

HeaderUtilityMenuPage.propTypes = propTypes;

export default HeaderUtilityMenuPage;
