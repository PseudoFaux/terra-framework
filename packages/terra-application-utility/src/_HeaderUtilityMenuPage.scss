:local {
  .menu-page {
    margin: 0;
    padding: 0;
    /* stylelint-disable selector-max-compound-selectors*/
    .list-item {
      list-style-type: none;
      cursor: pointer;
      font-size: var(--terra-application-header-menu-page-list-item-font-size, 1rem);
      padding-bottom: var(--terra-application-header-menu-page-list-item-padding-bottom, 0.2143rem);
      padding-left: var(--terra-application-header-menu-page-list-item-padding-left, 0.8571rem);
      padding-top: var(--terra-application-header-menu-page-list-item-padding-top, 0.2143rem);

      &:hover,
      &:focus {
        background-color: var(--terra-application-header-menu-page-list-item-hover-focus-color, #ebf6fd);
      }

      .chevron {
        height: var(--terra-application-utility-menu-page-chevron-height, 0.8571rem);
        min-height: var(--terra-application-utility-menu-page-height, 0.8571rem);
        min-width: var(--terra-application-utility-menu-page-chevron-width, 0.8571rem);
        width: var(--terra-application-utility-menu-page-chevron-width, 0.8571rem);
      }
    }
    /* stylelint-enable selector-max-compound-selectors */
  }
}
