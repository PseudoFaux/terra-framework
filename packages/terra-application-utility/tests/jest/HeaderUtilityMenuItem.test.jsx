import React from 'react';
import { UserData } from '../../src/ApplicationUtility';
import HeaderUtilityMenuItem from '../../src/header/_HeaderUtilityMenuItem';

describe('HeaderUtilityMenuItem', () => {
  const mockOnChange = jest.fn();

  it('should render a header utility menu item', () => {
    const wrapper = shallow(
      <HeaderUtilityMenuItem
        itemKey="key"
        title="title"
        content={<UserData />}
        contentLocation="footer"
        isSelected
        isSelctable
        hasChevron
        onChange={mockOnChange}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
