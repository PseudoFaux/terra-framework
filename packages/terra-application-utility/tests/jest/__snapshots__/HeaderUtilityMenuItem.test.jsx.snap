// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderUtilityMenuItem should render a header utility menu item 1`] = `
<button
  className="footer-item"
  isSelctable={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <UserData
    userDetail={null}
    userName={null}
  />
</button>
`;
